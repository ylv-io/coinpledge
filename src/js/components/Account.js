import React from 'react';
import { connect } from 'react-redux';
import SetUsernameForm from './SetUsername';

import { setUsername } from '../services/web3/user';
import { getTransactionReceipt } from '../services/web3/web3';
import { setUsername as setUsernameAction } from '../actions/web3';

class Account extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
    };
  }

  componentDidMount() {
  }

  componentDidUpdate(prevProps) {
  }

  handleSubmit = async ({ username }, { resetForm, setSubmitting, setStatus }) => {
    try {
      const { props } = this;

      const result = await setUsername(username);
      setSubmitting(false);
      resetForm();

      props.dispatch(setUsernameAction(username));

      const receipt = await getTransactionReceipt(result);
    } catch (e) {
      console.log(e);
      setSubmitting(false);
    }
  }

  render() {
    const {
      account,
      username,
    } = this.props;

    return (
      <section className="section">
        <div className="container">
          <h4 className="title is-3"><a target="_blank" rel="noopener noreferrer" href={`https://ropsten.etherscan.io/address/${account}`}>{account ? account.substring(0, 10) : ''}</a></h4>
          <hr />

          <div className="columns">
            <div className="column is-half">
              <div className="box">
                <h4 className="title is-4">Username</h4>
                <hr />
                { !username ? <SetUsernameForm handleSubmit={this.handleSubmit} /> : (
                  <span className="title is-4">
                    {
                      username
                    }
                  </span>
                )}
              </div>
            </div>
          </div>

        </div>
      </section>
    );
  }
}

const mapStateToProps = (state, props) => ({
  ...state.blockchain,
});

export default connect(mapStateToProps)(Account);
