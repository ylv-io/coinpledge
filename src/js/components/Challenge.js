import React from 'react';
import moment from 'moment';

import {
  Link,
} from 'react-router-dom';

import { shortAddress } from '../utils/web3';

export default class Challenge extends React.Component {
  constructor(props) {
    super(props);
    this.props = props;
  }

  render() {
    const {
      challenge,
      handleWin,
      handleLoss,
    } = this.props;

    return (
      <div className="card">
        <header className="card-header ">
          <p className={`card-header-title ${challenge.resolved ? challenge.successed ? 'has-background-success' : 'has-background-danger' : ''}`}>
            <span className={challenge.resolved ? 'has-text-white' : ''}>
              { challenge.resolved ? challenge.successed ? 'Win' : 'Loss' : 'Do It' }
            </span>
          </p>
        </header>
        <div className="card-content">
          <p className="is-size-5">
            <span>I pledge to </span>
            <strong className="is-size-4">
              &quot;
              {challenge.name}
              &quot;
            </strong>
            <span> before </span>
            <strong>{moment.unix(challenge.startDate).add(challenge.time, 's').format('DD MMM YYYY')}</strong>
            <span> by staking </span>
            <strong>
              {challenge.value}
              <span> ether.</span>
            </strong>
          </p>
          <small>
            <span>by </span>
            <Link to={`/${challenge.user}`}>{challenge.username}</Link>
            /
            <Link to={`/${challenge.mentor}`}>{challenge.mentorname}</Link>
          </small>
        </div>
        <footer className="card-footer">
          { challenge.canResolve && !challenge.isSubmitting
            && (
            <a
              href="#"
              className="card-footer-item has-text-success"
              onClick={handleWin}
            >
              Win
            </a>
            )
          }
          { challenge.canResolve && !challenge.isSubmitting
            && (
              <a
                href="#"
                className="card-footer-item has-text-danger"
                onClick={handleLoss}
              >
              Loss
              </a>
            )
          }
          { !challenge.isMentor
            && (
              <React.Fragment>
                <a
                  href={`https://twitter.com/intent/tweet?text=${encodeURIComponent(`I pledge to "${challenge.name}" before ${moment.unix(challenge.startDate).add(challenge.time, 's').format('DD MMM YYYY')} by staking ${challenge.value} ether at coinpledge.ylv.io.`)}&hashtags=coinpledge`}
                  target="_blank"
                  rel="noopener noreferrer"
                  className="card-footer-item"
                >
                  <span className="icon">
                    <i className="fab fa-twitter" />
                  </span>
                </a>
                <a
                  href={`https://www.facebook.com/sharer/sharer.php?u=coinpledge.ylv.io&quote=${encodeURIComponent(`I pledge to "${challenge.name}" before ${moment.unix(challenge.startDate).add(challenge.time, 's').format('DD MMM YYYY')} by staking ${challenge.value} ether at coinpledge.ylv.io.`)}`}
                  target="_blank"
                  rel="noopener noreferrer"
                  className="card-footer-item"
                >
                  <span className="icon">
                    <i className="fab fa-facebook" />
                  </span>
                </a>
              </React.Fragment>
            )
          }
        </footer>
      </div>
    );
  }
}
