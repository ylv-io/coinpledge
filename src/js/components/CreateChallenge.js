import React from 'react'
import { addExpense } from '../actions/challenges';
import { connect } from 'react-redux';
import { createChallenge } from '../services/web3/challenge';
import { withFormik } from 'formik';

const CreateChallengeForm = props => {
  const {
    values,
    touched,
    errors,
    isSubmitting,
    handleChange,
    handleBlur,
    handleSubmit,
    handleReset,
    dirty,
  } = props;
  return (
    <form onSubmit={handleSubmit}>
      <div className="field">
        <label className="label" htmlFor="name">State your goal. Keep it short.</label>
        <div className="control">
          <input 
            id="name"
            type="text" 
            name="name"
            value={values.name}
            onChange={handleChange}
            onBlur={handleBlur}
            className={
              errors.name && touched.name ? (
                'input is-danger'
              ) : (
                'input'
              )
            }
          />
          {errors.name && touched.name && (
              <div className="has-text-danger">{errors.name}</div>
            )}
        </div>
      </div>

      <div className="field">
        <label className="label" htmlFor="number">How much in ether? 0.1 is minimum.</label>
          <div className="field has-addons">
            <p className="control is-expanded">
              <input 
                type="number" 
                step="0.01" 
                name="value"
                id="value"
                value={values.value}
                onChange={handleChange}
                onBlur={handleBlur}
                className={
                  errors.value && touched.value ? (
                    'input is-danger'
                  ) : (
                    'input'
                  )
                }
              />
            </p>
            <p className="control">
              <a className="button is-static">
                ether
              </a>
            </p>
          </div>
          {errors.value && touched.value && (
            <div className="has-text-danger">{errors.value}</div>
          )}
      </div>

      <div className="field">
        <label className="label" htmlFor="mentor">Who will resolve challenge? Ethereum address required.</label>
        <div className="field">
          <p className="control">
            <input 
              type="text" 
              name="mentor"
              value={values.mentor}
              onChange={handleChange}
              onBlur={handleBlur}
              className={
                errors.mentor && touched.mentor ? (
                  'input is-danger'
                ) : (
                  'input'
                )
              }
            />
          </p>
          {errors.mentor && touched.mentor && (
            <div className="has-text-danger">{errors.mentor}</div>
          )}
        </div>
      </div>

      <div className="field">
        <label className="label" htmlFor="time">How long it will take in days?</label>
        <div className="field has-addons">
          <p className="control is-expanded">
            <input 
              type="number" 
              name="time"
              value={values.time}
              onChange={handleChange}
              onBlur={handleBlur}
              className={
                errors.time && touched.time ? (
                  'input is-danger'
                ) : (
                  'input'
                )
              }
            />
          </p>
          <p className="control">
            <a className="button is-static">
              days
            </a>
          </p>
        </div>
        {errors.time && touched.time && (
          <div className="has-text-danger">{errors.time}</div>
        )}
      </div>

      <hr/>
      <div className="field">
        <div className="control">
          <button type="submit" disabled={isSubmitting} className="button is-primary">Create Challenge</button>
        </div>
      </div>

    </form>
  );
};

const CreateChallengeWrappedForm = withFormik({
  mapPropsToValues: () => ({ 
    name: '',
    value: '',
    mentor: '',
    time: ''
  }),

  // Custom sync validation
  validate: values => {
    let errors = {};
    if(!values.name) {
      errors.name = 'Required';
    }
    if(!values.value || values.value <= 0) {
      errors.value = 'Required';
    }
    if(!values.mentor) {
      errors.mentor = 'Required';
    }
    if(!values.time || values.time <= 0) {
      errors.time = 'Required';
    }
    return errors;
  },

  handleSubmit: (values, { setSubmitting }) => {

    createChallenge(values.name, values.value, values.time, values.mentor)
      .then((result) => {
      })
      .catch(function(e) {
        
      });
    setSubmitting(false);
  },

  displayName: 'CreateChallengeForm', // helps with React DevTools
})(CreateChallengeForm);


class CreateChallenge extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isSuccessModelOpen: true
    }

  }
  render() {
    return (
      <section className="section">
        <div className="container">
          <div className="columns">
            <div className="column is-half box">
              <div>
                <h4 className="title is-4">Create Challenge</h4>
                <hr/>

                <CreateChallengeWrappedForm />
              
             </div>
            </div>
          </div>
        </div>

        <div className={
          this.state.isSuccessModelOpen ?
            'modal is-active' :
            'modal'
        }>
          <div 
            className="modal-background"
            onClick={() => {this.setState(() => ({ isSuccessModelOpen: false }))}}
          ></div>
          <div className="modal-content">
            <article className="message is-success">
              <div className="message-header">
                <p>Success</p>
                <button 
                  onClick={() => {this.setState(() => ({ isSuccessModelOpen: false }))}}
                  className="delete" 
                  aria-label="delete"
                />
              </div>
              <div className="message-body">
                <span>Your challenge has been submitted to Ethereum blockchain. it may take a while for transaction to be processed.</span>
              </div>
            </article>
          </div>
          <button 
            onClick={() => {this.setState(() => ({ isSuccessModelOpen: false }))}}
            className="modal-close is-large" 
            aria-label="close"
          />
        </div>

      </section>
    );
  }
}

export default connect()(CreateChallenge);