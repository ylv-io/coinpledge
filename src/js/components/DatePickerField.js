import React from 'react';

import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class DatePickerField extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      calendarFocused: false,
    };
  }

  handleChange = (value) => {
    const { props } = this;
    props.onChange('time', value);
  };

  onFocusChange = ({ focused }) => {
    const { props } = this;
    this.setState(() => ({ calendarFocused: focused }));
    if (!focused) props.onBlur('time', true);
  };

  render() {
    const { value, error, touched } = this.props;
    const { handleChange, onFocusChange } = this;
    const { calendarFocused } = this.state;

    return (
      <div className="field">
        <div className="notification">
          A date before which your goal has to be reached.
        </div>
        <label className="label" htmlFor="time">Deadline</label>
        <div className="control">
          <SingleDatePicker
            date={value}
            onDateChange={handleChange}
            focused={calendarFocused}
            onFocusChange={onFocusChange}
            numberOfMonths={1}
          />
        </div>
        {!!error
          && touched && (
            <div className="has-text-danger">{error}</div>
        )}
      </div>
    );
  }
}
