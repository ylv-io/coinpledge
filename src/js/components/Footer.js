import React from 'react';

const Footer = () => (
  <footer className="footer">
    <div className="content has-text-centered">
      <p>
        <a href="https://twitter.com/ylv_io">Twitter</a>
        <span> | </span>
        <a href="https://medium.com/@ylv">Medium </a>
        <span> | </span>
        <a href="https://github.com/ylv-io/coinpledge">Github</a>
        <span> | </span>
        <a href="https://t.me/ylv_public">Telegram</a>
      </p>
      <p>
        Donate
        <br />
        <small>0x4632F4120DC68F225e7d24d973Ee57478389e9Fd</small>
      </p>
      <p>
        <span>Built by </span>
        <a href="https://ylv.io/">Igor Yalovoy</a>
      </p>
    </div>
  </footer>
);

export default Footer;
