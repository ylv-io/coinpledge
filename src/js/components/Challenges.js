import React from 'react'
import { connect } from 'react-redux';
import Challenge from './Challenge';
import { resolveChallenge } from '../services/web3/challenge';
import { updateChallenge } from '../actions/challenges';

class Challenges extends React.Component {
  constructor(props) {
    super(props);
  }
  getHandleResolve = (challenge, decision) => {
    return async (e) => {
      e.preventDefault();
      const hash = await resolveChallenge(challenge.id, decision);
      this.props.dispatch(updateChallenge(challenge.id, { isSubmitting: true}));
    };
  }

  render() {
    return(
      <section className="section">
        <div className="container">
          <h4 className="title is-4">Your Challenges</h4>
          <hr/>
          { !this.props.challenges.length && <p className="title is-4">You don't have any challenges yet. Create one. You can do it!</p>}
          <div className="columns is-multiline">
            {this.props.challenges.map((o) => 
              <div className="column is-4" key={o.id}>
                <Challenge 
                  challenge={o} 
                  handleWin={this.getHandleResolve(o, true)} 
                  handleLoss={this.getHandleResolve(o, false)} 
                  />
              </div>)}
          </div>

          {!!this.props.history.length && (
            <div>
              <h4 className="title is-4">History</h4>
              <hr/>
              <div className="columns is-multiline">
                {this.props.history.map((o) => 
                  <div className="column is-4" key={o.id}>
                    <Challenge 
                      challenge={o} 
                      />
                  </div>)}
              </div>
            </div>
          )}

        </div>
      </section>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    challenges: state.challenges.filter(o => !o.resolved),
    history: state.challenges.filter(o => o.resolved)
  }
};

export default connect(mapStateToProps)(Challenges);