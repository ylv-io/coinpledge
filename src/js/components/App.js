import React from 'react';
import ReactDOM from 'react-dom';
import Web3 from 'web3';
import _ from 'underscore';

import CoinPledgeContract from '../../../build/contracts/CoinPledge.json';

import Challenge from './Challenge';
import CreateChallenge from './CreateChallenge';

import Contract from 'truffle-contract';


export default class App extends React.Component {

  state = {
    challenges: [],
    cases: [],
    bonusFund: 0
  }

  constructor(props){
    super(props)

    if(typeof web3 != 'undefined') {
        console.log("Using web3 detected from external source like Metamask")
        this.web3 = new Web3(web3.currentProvider);
        this.account = this.web3.eth.accounts[0];
    } else {
        console.log("No web3 detected.");
    }

    // Using truffle-contract we create the coinpledge object.
    const coin = Contract(CoinPledgeContract);
    coin.setProvider(this.web3.currentProvider);

    // Find contract instance on blockchain and bind
    coin.deployed().then((instance) => {
      this.state.coin = instance;

      this.updateState();
    });

  }

  getBonusFund = () => {
    const coin = this.state.coin;
    if(coin) {
      return coin.getBonusFund.call(this.account, {
        from: this.account
      })
      .then((result) => {
        const bonusFund = web3.fromWei(result.toNumber(), 'ether');
        console.log(`User has ${bonusFund} ether in bonus fund`);
        this.setState((o) => {
          return {
            bonusFund: bonusFund
          }
        });
      })
      .catch((e) => {
        console.log(e);
      });
    }
  }

  getAllCases = () => {
    const instance = this.state.coin;
    if(instance) {
      // check number of challenges
      instance.judgeToChallengeCount.call(web3.eth.accounts[0], {
        from: web3.eth.accounts[0]
      })
      .then((result) => {
        const numberOfCases = result.toNumber();
        console.log(`User has ${numberOfCases} cases`);
        if(numberOfCases > 0)
        // get all user cases indexes
          return instance.getCases.call(web3.eth.accounts[0], {
            from: web3.eth.accounts[0]
          });
        else return Promise.reject("User has zero cases");
      })
      .then((result) => {
        // get all users's cases objects
        const promises = result.map(((o) => instance.challenges.call(o.toNumber(), {
          from: web3.eth.accounts[0]
        })));

        // push indexes down the chain
        promises.push(result);
  
        return Promise.all(promises);
      })
      .then((result) => {
        //pop indexes
        const indexes = result.pop();
        // map all user's cases to objects
        const cases = result.map((o, i) => this.arrayToChallenge(o, indexes[i].toNumber()));
        console.log(`User cases:`);
        console.log(cases);
        this.setState((o) => {
          return {
            cases: cases
          }
        });
      })
      .catch(function(e) {
        console.log(e);
      });
    }
  }

  getAllChallenges = () => {
    const instance = this.state.coin;
    if(instance) {
      // check number of challenges
      instance.userToChallengeCount.call(web3.eth.accounts[0], {
        from: web3.eth.accounts[0]
      })
      .then((result) => {
        const numberOfChallenges = result.toNumber();
        console.log(`User has ${numberOfChallenges} challenges`);
        if(numberOfChallenges > 0)
        // get all user challenges indexes
          return instance.getChallenges.call(web3.eth.accounts[0], {
            from: web3.eth.accounts[0]
          });
        else return Promise.reject("User has zero challenges");
      })
      .then((result) => {
        // get all users's challenges objects
        const promises = result.map(((o) => instance.challenges.call(o.toNumber(), {
          from: web3.eth.accounts[0]
        })));

        // push indexes down the chain
        promises.push(result);
  
        return Promise.all(promises);
      })
      .then((result) => {
        //pop indexes
        const indexes = result.pop();
        // map all user's challenges to objects
        const challenges = result.map((o, i) => this.arrayToChallenge(o, indexes[i].toNumber()));
        console.log(`User challenges:`);
        console.log(challenges);
        this.setState((o) => {
          return {
            challenges: challenges
          }
        });
      })
      .catch(function(e) {
        console.log(e);
      });
    }
  }

  arrayToChallenge(array, id) {
    return {
      id: id,
      name: array[0],
      value: web3.fromWei(array[1].toNumber(), 'ether'),
      judge: array[2],
      startDate: array[3].toNumber(),
      time: array[4].toNumber(),
      successed: array[5],
      resolved: array[6],
      canResolve: this.account === array[2]
    }
  }

  componentDidMount() {
    setInterval(this.updateState.bind(this), 7e3);
  }

  updateState() {
    this.getAllChallenges();
    this.getBonusFund();
    this.getAllCases();
  }

  createChallenge = (e) => {
    e.preventDefault();

    const name = e.target.elements.name.value.trim();
    const value = e.target.elements.value.value.trim();
    const time = e.target.elements.time.value.trim();
    const judge = e.target.elements.judge.value.trim();

    this.state.coin.createChallenge(name, judge, time * 86400, {
        from: web3.eth.accounts[0],
        value: web3.toWei(value, 'ether')
      })
      .then((result) => {
        e.target.elements.name.value = '';
        e.target.elements.value.value = '';
        e.target.elements.time.value = '';
        e.target.elements.judge.value = '';
        console.log(result);
      })
      .catch(function(e) {
        console.log(e);
      });
  }

  getHandleResolve(id, decision) {
    return () => this.resolveChallenge(id, decision);
  }

  resolveChallenge = (id, decision) => {
    this.state.coin.resolveChallenge(id, decision, {
      from: web3.eth.accounts[0]
    })
    .then((result) => {
      console.log(result);
    })
    .catch(function(e) {
      console.log(e);
    });
  }

  render () {
    return (
        <div>
          <section className="hero is-success is-bold">
            <div className="hero-body">
              <div className="container">
                <h1 className="title">CoinPledge</h1>
                <h4 className="subtitle">Archive your goals and connect with people in a meaningful way</h4>
                <hr/>
              </div>
            </div>
          </section>

          <section className="section">
            <div className="container">
              <p>
                  Your bonus fund has <strong>{this.state.bonusFund}</strong> ether. <br/>
                  Your account is <strong>{this.account}</strong>.
              </p>
            </div>
          </section>

          <section className="section">
            <div className="container">
              <div className="columns">
                <div className="column is-half box">
                  <CreateChallenge createChallenge={this.createChallenge} />
                </div>
              </div>
            </div>
          </section>
          <section className="section">
            <div className="container">
              <h4 className="title is-4">Your Challenges</h4>
              <hr/>
              { !this.state.challenges.length && <p className="title is-4">You don't have any challenges yet. Create one. You can do it!</p>}
              <div className="columns is-multiline">
                {this.state.challenges.map((o) => 
                  <div className="column is-4" key={o.id}>
                    <Challenge challenge={o} handleWin={this.getHandleResolve(o.id, true)} handleLoss={this.getHandleResolve(o.id, false)} />
                  </div>)}
              </div>
            </div>
          </section>
          <section className="section">
            <div className="container">
            <h4 className="title is-4">Your Cases</h4>
            <hr/>
            { !this.state.challenges.length && <p className="title is-4">You don't have any cases yet. Help someone!</p>}
              <div className="columns is-multiline">
                {this.state.cases.map((o) => 
                  <div className="column is-4" key={o.id}>
                    <Challenge challenge={o} handleWin={this.getHandleResolve(o.id, true)} handleLoss={this.getHandleResolve(o.id, false)} />
                  </div>
                )}
              </div>
            </div>
          </section>

        </div>
    )
  }
}
