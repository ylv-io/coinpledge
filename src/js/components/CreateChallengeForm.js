import React from 'react';
import { connect } from 'react-redux';
import { withFormik } from 'formik';

import moment from 'moment';

import DatePickerField from './DatePickerField';
import SelectField from './SelectField';

class CreateChallengeForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
    };
  }

  render() {
    const {
      values,
      touched,
      errors,
      isSubmitting,
      handleChange,
      handleBlur,
      handleSubmit,
      handleReset,
      setFieldValue,
      setFieldTouched,
      dirty,
      users,
    } = this.props;
    return (
      <form onSubmit={handleSubmit}>
        <div className="field">
          <div className="notification">
            Goal has to be specific and measurable. Good example &quot;100 pull ups&quot;, bad example &quot;be more healthy&quot;. Keep it short.
          </div>
          <label className="label" htmlFor="name">Goal</label>
          <div className="control">
            <input
              id="name"
              type="text"
              name="name"
              value={values.name}
              onChange={handleChange}
              onBlur={handleBlur}
              className={
                errors.name && touched.name ? (
                  'input is-danger'
                ) : (
                  'input'
                )
              }
            />
            {
              errors.name && touched.name && (
                <div className="has-text-danger">{errors.name}</div>
              )
          }
          </div>
        </div>

        <hr />

        <div className="field">
          <div className="notification">
            Amount to stake. 7 cappuccinos is a good stake.
          </div>
          <label className="label" htmlFor="value">Stake</label>
          <div className="field has-addons">
            <p className="control is-expanded">
              <input
                type="number"
                step="0.01"
                name="value"
                id="value"
                value={values.value}
                onChange={handleChange}
                onBlur={handleBlur}
                className={
                    errors.value && touched.value ? (
                      'input is-danger'
                    ) : (
                      'input'
                    )
                  }
              />
            </p>
            <p className="control">
              <a className="button is-static">
                  ether
              </a>
            </p>
          </div>
          {errors.value && touched.value && (
          <div className="has-text-danger">{errors.value}</div>
          )}
        </div>

        <hr />
        <SelectField
          value={values.mentor}
          onChange={setFieldValue}
          onBlur={setFieldTouched}
          error={errors.mentor}
          touched={touched.touched}
          options={users}
        />

        <hr />
        <div className="field">
          <div className="notification">
            Paid out of the staked amount. There is 10% service fee.
          </div>
          <label className="label" htmlFor="mentorFee">Mentor Reward</label>
          <div className="field has-addons">
            <p className="control is-expanded">
              <input
                type="number"
                step="0.01"
                name="mentorFee"
                id="mentorFee"
                value={values.mentorFee}
                onChange={handleChange}
                onBlur={handleBlur}
                className={
                    errors.mentorFee && touched.mentorFee ? (
                      'input is-danger'
                    ) : (
                      'input'
                    )
                  }
              />
            </p>
            <p className="control">
              <a className="button is-static">
                  ether
              </a>
            </p>
          </div>
          {errors.mentorFee && touched.mentorFee && (
          <div className="has-text-danger">{errors.mentorFee}</div>
          )}
        </div>

        <hr />
        <DatePickerField
          value={values.time}
          onChange={setFieldValue}
          onBlur={setFieldTouched}
          error={errors.time}
          touched={touched.time}
        />

        <hr />
        <div className="field">
          <div className="control">
            <button type="submit" disabled={isSubmitting} className="button is-primary is-fullwidth">Do It</button>
          </div>
        </div>

      </form>
    );
  }
}

export default withFormik({
  mapPropsToValues: () => ({
    name: '',
    value: '',
    mentor: '',
    mentorFee: '',
    time: moment().add(7, 'days'),
  }),

  // Custom sync validation
  validate: (values) => {
    const errors = {};
    if (!values.name) {
      errors.name = 'Required';
    }

    if (values.value <= 0) {
      errors.value = 'Required';
    } else if (values.value < 0.1) { errors.value = 'Mininum amount is 0.1 eth'; }

    if (!values.mentor) {
      errors.mentor = 'Required';
    }

    if (values.mentorFee === '' || values.mentorFee < 0) {
      errors.mentorFee = 'Required';
    } else if (values.mentorFee > values.value) {
      errors.mentorFee = 'Should be equal or less than stake';
    }

    if (!values.time || values.time <= 0) {
      errors.time = 'Required';
    }
    return errors;
  },

  handleSubmit: (values, bag) => {
    bag.props.handleSubmit(values, bag);
  },

  displayName: 'CreateChallengeForm', // helps with React DevTools
})(CreateChallengeForm);
