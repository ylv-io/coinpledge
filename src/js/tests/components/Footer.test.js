import React from 'react';
import { shallow } from 'enzyme';
import { Footer } from '../../components/Footer';

test('should render footer correctly', () => {
  const wrapper = shallow(<Footer donations={[]} />);
  expect(wrapper).toMatchSnapshot();
});
