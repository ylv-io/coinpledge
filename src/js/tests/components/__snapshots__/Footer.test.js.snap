// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render footer correctly 1`] = `
<Fragment>
  <footer
    className="footer"
  >
    <div
      className="content has-text-centered"
    >
      <div
        className="columns"
      >
        <div
          className="column is-offset-one-quarter is-half"
        />
      </div>
      <p>
        <a
          href="https://twitter.com/ylv_io"
          rel="noopener noreferrer"
          target="_blank"
        >
          Twitter
        </a>
        <span>
           | 
        </span>
        <a
          href="https://medium.com/@ylv"
          rel="noopener noreferrer"
          target="_blank"
        >
          Medium 
        </a>
        <span>
           | 
        </span>
        <a
          href="https://github.com/ylv-io/coinpledge"
          rel="noopener noreferrer"
          target="_blank"
        >
          Github
        </a>
        <span>
           | 
        </span>
        <a
          href="https://t.me/ylv_public"
          rel="noopener noreferrer"
          target="_blank"
        >
          Telegram
        </a>
        <span>
           | 
        </span>
        <a
          href="https://etherscan.io/address/0x896F11b2628208e09BD5CFdD0C4f7C28C0349d61"
          rel="noopener noreferrer"
          target="_blank"
        >
          Etherscan
        </a>
        <span>
           | 
        </span>
        <a
          href="https://www.stateofthedapps.com/dapps/coinpledge"
          rel="noopener noreferrer"
          target="_blank"
        >
          State of dApps
        </a>
      </p>
      <p>
        <span>
          Built by 
        </span>
        <a
          href="https://ylv.io/"
        >
          Igor Yalovoy
        </a>
      </p>
    </div>
  </footer>
</Fragment>
`;
